### PR TITLE
Convert temperature challenge

### PR DESCRIPTION
**Pull request by Shaymamohd**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_celsius_to_fahrenheit.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        """It should handle large temperatures correctly"""
+        self.assertEqual(celsius_to_fahrenheit(1000), 1832.0)
+
+    def test_very_small_fraction(self):
+        """It should correctly convert very small fractional Celsius values"""
+        self.assertAlmostEqual(celsius_to_fahrenheit(0.0001), 32.00018)
+
+    def test_extreme_negative(self):
+        """It should handle extremely large negative temperatures"""
+        self.assertEqual(celsius_to_fahrenheit(-1000000), -1799968.0)
```
Comment (Adamx090) [inline solutions/tests/test_celsius_to_fahrenheit.py:53]: @Shaymamohd Excellent, very organized and well thought out.
I just have one comment here about the extreme negative case. The lowest possible temperature is absolute zero Kelvin, which is -273.15 degrees Celsius, it doesn't go lower than this. So I suggest you remove this extreme case from the unittests, and add an assertion in the code.
`assert celsius >= -273.15, "Temperature can't be below absolute zero"` or something like this, and then add an additional defensive test for this assertion.

**Review by Adamx090 — CHANGES_REQUESTED** [review]

File: solutions/tests/test_celsius_to_fahrenheit.py (hunk 1 above)
Comment (Shaymamohd) [reply solutions/tests/test_celsius_to_fahrenheit.py]: Thank you, Adam!
I just learned about absolute zero from your comment—really appreciate the info! I’ve made the changes based on your review, including removing the extreme negative case and adding the necessary assertion. Thanks again for your help!
you can review it again now

File: solutions/tests/test_celsius_to_fahrenheit.py [hunk 2]
```diff
@@ -48,9 +48,9 @@ def test_very_small_fraction(self):
         """It should correctly convert very small fractional Celsius values"""
         self.assertAlmostEqual(celsius_to_fahrenheit(0.0001), 32.00018)
 
-    def test_extreme_negative(self):
-        """It should handle extremely large negative temperatures"""
-        self.assertEqual(celsius_to_fahrenheit(-1000000), -1799968.0)
+    def test_absolute_zero(self):
+        """Test the conversion at absolute zero."""
+        self.assertEqual(celsius_to_fahrenheit(-273.15), -459.67)
```
Comment (Adamx090) [inline solutions/tests/test_celsius_to_fahrenheit.py]: CI checks failed this test due to this rounding issue (-459.66999999999996 != -459.67). I don't know why this happened. The formula divides by 5, so it should be an exact -459.67. But a quick fix can be like this:
`fahrenheit = round((celsius * 9 / 5) + 32), 2)`

**Review by Adamx090 — CHANGES_REQUESTED** [review]

File: solutions/tests/test_celsius_to_fahrenheit.py (hunk 2 above)
Comment (Shaymamohd) [reply solutions/tests/test_celsius_to_fahrenheit.py]: I made two commits, and the second commit addressed the issue in the first one.
It seems like you reviewed the first commit, but I made edits in the second one. 
In the second commit, I changed the test from:
   self.assertEqual(celsius_to_fahrenheit(-273.15), -459.67)
   to 
   self.assertAlmostEqual(celsius_to_fahrenheit(-273.15), -459.67) and it passed the checks

Comment (Adamx090) [reply solutions/tests/test_celsius_to_fahrenheit.py]: Great job! Merging Now..

**Review by Adamx090 — APPROVED** [review]
